### PR TITLE
Added a proper working revalidateTag functionality to redis-stack handler

### PR DESCRIPTION
**Pull request by MaKTaiL**
"revalidateTag" wasn't working correctly because it wasn't deleting the keys associated with the tag.

- Added index creation for searching tags accross multiple keys
- Added function for sanitizing tag string (to avoid errors with index search)
- revalidateTag finds and deletes all keys with the requested tag

### CODE REVIEW

**Comment by better-salmon** [comment]
Hi @MaKTaiL, thank you for your contribution. The current implementation is fine. Instead of deleting cache entries, we are now collecting a list of tags that have been revalidated. When the Next.js Server retrieves values from the cache store, we check if a particular cache key was revalidated before serving it to the Next.js server. This was done to support pages from Pages Router. I'm currently working on a major refactoring for this part of the library. Once I've completed that, I'll revisit your PR. Thanks again!

**Comment by MaKTaiL** [comment]
> Hi @MaKTaiL, thank you for your contribution. The current implementation is fine. Instead of deleting cache entries, we are now collecting a list of tags that have been revalidated. When the Next.js Server retrieves values from the cache store, we check if a particular cache key was revalidated before serving it to the Next.js server. This was done to support pages from Pages Router. I'm currently working on a major refactoring for this part of the library. Once I've completed that, I'll revisit your PR. Thanks again!

The current implementation does not do anything when you request a revalidateTag, it only updates the time the tag was "updated". The cache entry values remain the same unless you delete the keys associated with the tag from the Redis database. Deleting the keys will signal NextJS that there is no cache entry for it which will force a revalidation of the fetch results which will then be saved again with new values on Redis. You can easily check this behavior by enabling the logs from this library:

https://github.com/alfonsusac/nextjs-better-unstable-cache

I also made a function for sanitizing tag strings (Redis only allows a-z A-Z 0-9 for query/get/set operations).
And finally I created an index for easily querying all keys that contain the requested tag for revalidation.

**Comment by better-salmon** [comment]
Hey @MaKTaiL! Sorry for the delay. Please change the target branch of this PR to the [canary-1.0.0-rc](https://github.com/caching-tools/next-shared-cache/tree/canary-1.0.0-rc).

**Comment by MaKTaiL** [comment]
Hi @better-salmon! I just changed target branch, merged and resolved conflicts. I saw the changes you've made and it appears to be a working revalidateTag example but I'm concerned about the race condition of storing the tags in a map. If multiple people request different tags for revalidation, the tags' map (stored in a key formed with keyPrefix + sharedTagsKey) has a huge chance of conflicting between requests since both will be updating it at the same time. Whichever updates last will be the winner.

Let's look at this scenario:

Tags' map looks like this:

``` json
{
  "tag1": ["key1", "key2"],
  "tag2": ["key3', "key4"],
  "tag3": ["key5', "key6"],
}
```

- User 1 requests revalidation of tag1, and User 2 requests revalidation of tag2, both at the same time.
- Both requests retrieve the same copy of the tags map
- First request removes tag1 from map and sends it back to Redis.
- Second request removes tag2 and sends it back to Redis.
- Final map will look like this:

``` json
{
  "tag1": ["key1", "key2"],
  "tag3": ["key5', "key6"],
}
```

You see tag1 still remains because it wasn't aware of the User 1 request. Now imagine a scenario with thousand of people doing the same thing. There is also the race condition of multiple keys attached to the same tag, updating the map at the same time.
That said I think using an index for the tags is the better, faster and safer solution.
For using an index it is paramount that we sanitize the tags (remove any chars that Redis does not accept in query operations).
I adapted the code for this while also using the sharedTagsKey you implemented. Let me know what you think.

**Comment by better-salmon** [comment]
I oversaw this case during the major refactoring. Thank you! I'll review and merge your solution soon.

**Comment by better-salmon** [comment]
@MaKTaiL hi! I have ended the work on the new API. But I accidentally made some force pushes to the `canary-1.0.0-rc` branch, please rebase your PR one more time. I'm sorry.

And I want you to make some fixes:

- Run npm run codestyle:fix before committing.
- Use full names for variables and arguments e.g. `query.documents.map((doc) => doc.id)` will become `query.documents.map((document) => document.id)` or `query.documents.map(({ id }) => id)`
- Remove the sharedTagsKey option from the redis-stack Handler options. It seems obsolete with indexing.

**Comment by better-salmon** [comment]
Actually, I can merge `canary-1.0.0-rc` to `canary`, and it will be easier for you to merge your PR to the `canary` branch.

**Comment by better-salmon** [comment]
Closing your pull request was not intended. Please rebase it one last time to the `canary` with the requested changes. You've done a great job!

**Comment by MaKTaiL** [comment]
Hi @better-salmon , sorry for the late reply. I was busy this weekend. I'll merge the changes as soon as I can.

**Comment by better-salmon** [comment]
It's ok! I will wait 👍

**Comment by MaKTaiL** [comment]
@better-salmon All good! I've done what you asked:

- Run npm run codestyle:fix before committing.
- Use full names for variables and arguments e.g. query.documents.map((doc) => doc.id) will become query.documents.map((document) => document.id) or query.documents.map(({ id }) => id)
- Remove the sharedTagsKey option from the redis-stack Handler options. It seems obsolete with indexing.

File: packages/cache-handler/src/handlers/server.ts [hunk 1]
```diff
@@ -65,6 +65,8 @@ export default function createHandler({ baseUrl, timeoutMs }: ServerCacheHandler
         async set(key, cacheHandlerValue) {
             const url = new URL('/set', baseUrl);
 
+            const _isRouteKind = cacheHandlerValue.value?.kind === 'ROUTE';
```
Comment (better-salmon) [inline packages/cache-handler/src/handlers/server.ts]: Delete this line please

File: packages/cache-handler/src/handlers/redis-stack.ts [hunk 2]
```diff
@@ -55,7 +54,6 @@ export type RedisCacheHandlerOptions<T> = {
 export default async function createHandler<T extends RedisClientType>({
     client,
     keyPrefix = '',
-    sharedTagsKey = '__sharedTags__',
     timeoutMs = 5000,
 }: RedisCacheHandlerOptions<T>): Promise<Handler> {
```
Comment (better-salmon) [inline packages/cache-handler/src/handlers/redis-stack.ts]:
```suggestion
}: Omit<RedisCacheHandlerOptions<T>, 'sharedTagsKey'>): Promise<Handler> {
```

File: packages/cache-handler/src/handlers/redis-stack.ts [hunk 3]
```diff
@@ -79,61 +97,43 @@ export default async function createHandler<T extends RedisClientType>({
...
-                  })
-                : undefined;
+            if (cacheHandlerValue.value?.kind === 'ROUTE') {
+                preparedCacheValue = structuredClone(cacheHandlerValue);
+                // @ts-expect-error -- object must have the same shape as cacheValue
+                preparedCacheValue.value.body = cacheHandlerValue.value.body.toString('base64') as unknown as Buffer;
+            }
+
+            preparedCacheValue.tags = preparedCacheValue.tags.map(sanitizeTag);
+
```
Comment (better-salmon) [inline packages/cache-handler/src/handlers/redis-stack.ts]:
```suggestion
            cacheHandlerValue.tags = cacheHandlerValue.tags.map(sanitizeTag);

```

File: packages/cache-handler/src/handlers/redis-stack.ts [hunk 4]
```diff
@@ -79,61 +97,43 @@ export default async function createHandler<T extends RedisClientType>({
...
+            const options = getTimeoutRedisCommandOptions(timeoutMs);
 
             const setCacheValue = client.json.set(
                 options,
                 keyPrefix + key,
                 '.',
-                cacheHandlerValue as unknown as RedisJSON,
+                preparedCacheValue as unknown as RedisJSON,
```
Comment (better-salmon) [inline packages/cache-handler/src/handlers/redis-stack.ts]:
```suggestion
                cacheHandlerValue as unknown as RedisJSON,
```

File: packages/cache-handler/src/handlers/redis-stack.ts [hunk 5]
```diff
@@ -79,61 +97,43 @@ export default async function createHandler<T extends RedisClientType>({
...
             );
 
-            const expireCacheValue = cacheHandlerValue.lifespan
-                ? client.expireAt(options, keyPrefix + key, cacheHandlerValue.lifespan.expireAt)
+            const expireCacheValue = preparedCacheValue.lifespan
+                ? client.expireAt(options, keyPrefix + key, preparedCacheValue.lifespan.expireAt)
                 : undefined;
```
Comment (better-salmon) [inline packages/cache-handler/src/handlers/redis-stack.ts]:
```suggestion
            const expireCacheValue = cacheHandlerValue.lifespan
                ? client.expireAt(options, keyPrefix + key, cacheHandlerValue.lifespan.expireAt)
                : undefined;

```

**Review by better-salmon — CHANGES_REQUESTED** [review]
Please, remove some outdated code I've commented

File: packages/cache-handler/src/handlers/redis-stack.ts (hunk 3 above)
Comment (better-salmon) [reply packages/cache-handler/src/handlers/redis-stack.ts]: I'm sorry, suggestion is confusing. I meant that there is no need in creating `preparedCacheValue` and checking for the `ROUTE` kind anymore.

**Comment by MaKTaiL** [comment]
@better-salmon For some reason it wasn't letting me apply your suggestions here so I manually edited them and pushed the changes. I think that is all. Let me know if there are any more fixes to apply.

File: packages/cache-handler/src/handlers/redis-stack.ts [hunk 6]
```diff
@@ -98,42 +112,20 @@ export default async function createHandler<T extends RedisClientType>({
...
-                if (tags.includes(tag)) {
-                    keysToDelete.push(keyPrefix + key);
-                    tagsToDelete.push(key);
-                }
-            }
-
-            if (keysToDelete.length === 0) {
-                return;
-            }
+            const keysToDelete = query.documents.map((document) => document.id);
```
Comment (better-salmon) [inline packages/cache-handler/src/handlers/redis-stack.ts:123]: Let's add check for `keysToDelete.length`. If there is no keys to revalidate there is no need to call `client.del`.

```js
if (keysToDelete.length === 0) {
    return;
}
```

**Review by better-salmon — CHANGES_REQUESTED** [review]

**Review by better-salmon — APPROVED** [review]

**Comment by better-salmon** [comment]
Thank you for your contribution!

**Pull request merged**